### DISCUSSION TITLE
Composite w/ CopyAlpha

### DISCUSSION THREAD

**Discussion by forensicmike** [dlemstra/Magick.NET#1414] [Help]
Hi all, thanks for taking the time to look. I am striking out on masking an image using the Composite / CopyAlpha operator.

I essentially have the same problem as this ancient issue: https://github.com/dlemstra/Magick.NET/issues/20

I've confirmed that the mask looks like it should - e.g. a black background with white pixels where I want pixels to be copied. I also confirmed that the mask's `HasAlpha` is set to false, as per the resolution for issue 20.

The output, however, is the entire input image. Doesn't matter what I do it's the **entire** input image!

I've tried using separate input images, including different formats (png and jpeg), and it simply does nothing.

Here's the code - I am using imagemagick.net q16-anyCPU v13.1.3. Note my use case is polygons but I've just gone with the circle here as it's what was asked in the referenced issue above, and is simpler code.

```csharp
using var mg = new MagickImage(CurrentFilePath);
//mg.VirtualPixelMethod = VirtualPixelMethod.Transparent;
using var mask = new MagickImage(MagickColors.Black, mg.Width, mg.Height);
mask.HasAlpha = false;

var diameter = mask.Width < mask.Height ? mask.Width : mask.Height;
int originY = (mask.Height - diameter) / 2 + diameter / 2;
int originX = (mask.Width - diameter) / 2 + diameter / 2;

var drawables = new Drawables();
drawables.FillColor(MagickColors.White);
drawables.StrokeColor(MagickColors.White);
drawables.Circle(originX, originY, 0, originY);

mask.Draw(drawables);

mg.Composite(mask, CompositeOperator.CopyAlpha);
using var ms = new MemoryStream();
mg.Write(ms, MagickFormat.Png);

ms.Position = 0;
BitmapImage bitmapImage = new();
bitmapImage.BeginInit();
bitmapImage.StreamSource = ms;
bitmapImage.CacheOption = BitmapCacheOption.OnLoad;
bitmapImage.EndInit();

Clipboard.SetImage(bitmapImage);
```

So to reiterate when I set `mg.Write(ms, MagickFormat.Png);` to `mask.Write(ms, MagickFormat.Png);` the result looks good:

![image](https://github.com/dlemstra/Magick.NET/assets/44839768/84b9b332-970f-47ae-b121-f4b2af45a01a)

But when I change it back to mg.Write, the pasted image looks identical to the input image - no masking at all.

Does anyone have any idea what I'm doing wrong?

I've also tried mg.Write with MagickFormat.Jpg just in case. Also, the mg.VirtualPixelMethod is commented out but I have tried it with and without and there is no difference in output. I've also tried pasting into multiple apps (Paint, Photoshop) with same result.

**Comment by forensicmike** [comment]
It looks like it's the Clipboards fault? I threw in a `File.WriteAllBytes(@"c:\temp\test.png", mg.ToByteArray());` and the resultant image is perfect. Closing this.

**Comment by dlemstra** [comment]
What happens when you use `ToBitmapSource` extension method from the `Magick.NET.SystemWindowsMedia` NuGet package?
